R/helpers-ppc.R

```diff
@@ -250,6 +276,203 @@ all_counts <- function(x, ...) {
+
+    # pre-compute quantiles and use symmetry for increased efficiency.
+    x2_lower <- qbinom(gamma / 2, N, z2)
+    x2_upper <- c(N - rev(x2_lower)[2:K], 1)
+
+    # Compute the total probability of trajectories inside the confidence
+    # intervals. Initialize the set and corresponding probabilities known
+    # to be 0 and 1 for the starting value z1 = 0.
+    x1 <- 0
+    p_int <- 1
+    for (i in seq_along(z1)) {
+      tmp <- p_interior(
+        p_int, x1 = x1, x2 = x2_lower[i]: x2_upper[i],
+        z1 = z1[i], z2 = z2[i], N = N
+      )
+      x1 <- tmp$x1
+      p_int <- tmp$p_int
+    }
+    abs(prob - sum(p_int))
+  }
+  optimize(target, c(0, 1 - prob), prob, N = N, K = K)$minimum
+}
+
+# Adjust coverage parameter for multiple chains using simulation method.
+adjust_gamma_simulate <- function(N, L, K, prob, M) {
+  gamma <- numeric(M)
+  z <- (1:(K - 1)) / K
+  n <- N * (L - 1)
+  k <- floor(z * N * L)
+  for (m in seq_len(M)) {
+    u <- u_scale(replicate(L, runif(N)))
+    scaled_ecdfs <- apply(outer(u, z, "<="), c(2, 3), sum)
+    gamma[m] <- 2 * min(
+      apply(
+        scaled_ecdfs, 1, phyper, m = N, n = n, k = k
+      ),
+      apply(
+        scaled_ecdfs - 1, 1, phyper, m = N, n = n, k = k, lower.tail = FALSE
+      )
+    )
+  }
+  alpha_quantile(gamma, 1 - prob)
+}
+
+interpolate_gamma <- function(N, K, p, L) {
+  vals <- get_interpolation_values(N, K, L, p)
+  N_lb <- max(vals[vals$N <= N, ]$N)
+  N_ub <- min(vals[vals$N >= N, ]$N)
+  g_lb <- approx(
+    x = log(vals[vals$N == N_lb, ]$K),
+    y = log(vals[vals$N == N_lb, ]$val),
+    xout = log(K)
+  )$y
+  g_ub <- approx(
+    x = log(vals[vals$N == N_ub, ]$K),
+    y = log(vals[vals$N == N_ub, ]$val),
+    xout = log(K)
+  )$y
+  if (N_ub == N_lb) {
+    g <- exp(g_lb)
+  } else {
+    g <- exp(approx(x = log(c(N_lb, N_ub)), y = c(g_lb, g_ub), xout = log(N))$y)
+  }
+  g
+}
+
+get_interpolation_values <- function(N, K, L, p) {
+  for (dim in c("L", "prob")) {
+    if (all(get(if (dim == "L") dim else "p") != bayesplot:::gamma_adj[, dim])) {
+      stop(paste(
+        "No precomputed values to interpolate from for '", dim, "' = ",
+        get(if (dim == "L") dim else "p"),
+        ".\n",
+        "Values of '", dim, "' available for interpolation: ",
+        unique(bayesplot:::gamma_adj[, dim]),
+        ".",
+        sep = ""
+      ))
+    }
+  }
+  vals <- bayesplot:::gamma_adj[bayesplot:::gamma_adj$L == L & bayesplot:::gamma_adj$prob == p, ]
+  if (K > max(vals[vals$N <= N, ]$K)) {
+    stop(paste(
+      "No precomputed values available for interpolation for 'K' = ",
+      K,
+      ".\n",
+      "Try either setting a value of 'K' <= ",
+      max(vals[vals$N <= N, ]$K),
+      "or 'interpolate_adj' = FALSE."
+    ))
+  }
+  vals
+}
+#' A helper function for 'adjust_gamma_optimize' defining the probability that
+#' an ECDF stays within the supplied bounds between z1 and z2.
+#' @noRd
+p_interior <- function(p_int, x1, x2, z1, z2, N) {
```

Comment:
Since the arguments are not informatively named, I would recommend adding some documentation on their meaning.